### ISSUE TITLE
Rubocop doesn't know the new lonely operator (CRuby 2.3.0)

### ISSUE THREAD

**Issue by Dbz** [bbatsov/rubocop#2713]
```
$ rubocop -V
0.36.0 (using Parser 2.3.0.1, running on ruby 2.3.0 x86_64-darwin15)
```

Ruby Code:

```ruby
class Node
  attr_accessor :left, :right, :data
  def to_s
    "(Data: #{@data}) : (Left: #{@left&.data}) : (Right: #{@right&.data})"
  end
end
```

Error message: 

```
E: unexpected token error
(Using Ruby 2.0 parser; configure using TargetRubyVersion parameter, under AllCops)
    "(Data: #{@data}) : (Left: #{@left&.data}) : (Right: #{@right&.data})"
                                                                 ^^
```

I would like to submit a pull request and fix this issue myself. Can I have some guidance on where to start?

**Comment by segiddins** [comment]
Set your target ruby version to 2.3 in your rubocop config

**Comment by Dbz** [comment]
There goes my chance to contribute! Perhaps I can add a script to modify the config if there is a .ruby-version file?

**Comment by alexdowad** [comment]
> Perhaps that value should be ||ed with the value from the .ruby-version file when it gets used?

That's been suggested before. It's an intriguing idea.

So what happens if someone has an explicit Ruby version in `.rubocop.yml`, and a different Ruby version in `.ruby-version`?

**Comment by Dbz** [comment]
There could be a boolean config variable right above or below `TargetRubyVersion` called something along the lines of `SetTargetRubyVersionFromCodebase` with a comment explaining it checks `.ruby-version` or the `Gemfile` to override `TargetRubyVersion`.

**Comment by alexdowad** [comment]
Another option would be to print a warning if `.ruby-version` does not match the configured `TargetRubyVersion`.

**Comment by alexdowad** [comment]
The message comes from `sym_check_asciionly`.

I'm just compiling a patched Ruby which prints the name of the invalid encoding, rather than just saying `invalid symbol encoding`.

**Comment by Dbz** [comment]
That's also a good idea. Is there a reason why the default value for `TargetRubyVersion` is `2.0.0`? Would it also make sense to set that to `2.3.0`?

**Comment by alexdowad** [comment]
> Is there a reason why the default value for TargetRubyVersion is 2.0.0?

So RuboCop's auto-correction doesn't break people's code by using new syntax.

(If you are developing a library, you probably want it to work on all Ruby versions which are still supported by the Ruby core team. If auto-correction changed your code to use 2.3 syntax, that could inadvertently break compatibility with older Rubies.)

**Comment by bbatsov** [comment]
I'm closing this ticket as its original purpose was fulfilled.

**Comment by Dbz** [comment]
Sounds fair. In the meantime I'd like to implement the warning when your `TargetRubyVersion` doesn't match what's in `.ruby-version`. Should I make a new cop? Or is there a better place to put this?

**Comment by alexdowad** [comment]
> Should I make a new cop? Or is there a better place to put this?

Cops find problems with *the inspected code*. This is a problem with someone's *RuboCop configuration*. So a new cop is not what you want.

We already have assorted validation logic for RC config in `Config`. This check should probably be done in `Config`, or else in `ConfigLoader` or a related class.

One thing to be aware of is that `Config` is instantiated, not just once, but once *for each inspected file*. So for performance, I suggest that whatever you do, you should not search for and read `.ruby-version` every time a `Config` object is created. Find some way to do that only once, whether it means caching the value, or in some other way.
